src/core/storage/PageCache.cpp

```diff
@@ -0,0 +1,63 @@
+#include <algorithm>
+#include <array>
+#include <fstream>
+#include <filesystem>
+
+#include <third-party/catch2/Catch2.h>
+
+#include <core/storage/PageCache.h>
+#include <core/storage/Page.h>
+#include <core/storage/Position.h>
+
+using namespace internal::storage;
+
+static constexpr std::string_view test_output_fname = "/tmp/eu-bufpool-test1";
+static constexpr std::size_t test_num_pages = 3;
+
+TEST_CASE("buffer-pool-ops", "[bufpool]") {
+	PageCache pgcache(test_output_fname, PageCache::min_size());
+	REQUIRE(!pgcache.full());
+
+	std::array<uint8_t, Page::size()> data;
+	auto make_dirty_page = [&, val = 0]() mutable {
+		auto p = Page::empty();
+		std::fill(data.begin(), data.end(), val++);
+		p.write(0, data.cbegin(), data.cend());
+
+		auto pos = pgcache.get_new_pos();
+		// Don't copy in order to be able to compare read page later
+		pgcache.put_page(pos, Page(p));
+		REQUIRE(pgcache.get_page(pos) == p);
+
+		auto &cpage = pgcache.get_page(pos);
+		std::fill(data.begin(), data.end(), val++);
+		cpage.write(0, data.cbegin(), data.cend());
+		pgcache.put_page(pos, Page(cpage));
+		REQUIRE(cpage == pgcache.get_page(pos));
+	};
+
+	for (size_t i = 0; i < test_num_pages; ++i) {
+		make_dirty_page();
+		REQUIRE(pgcache.full());
+	}
+
+	pgcache.flush_all();
+
+	char *ptr = reinterpret_cast<char *>(const_cast<unsigned char *>(data.data()));
+
+	REQUIRE(std::filesystem::file_size("/tmp/eu-bufpool-test1") == test_num_pages * Page::size());
+
+	std::fstream fbufpool("/tmp/eu-bufpool-test1", std::ios::in);
+	fbufpool.seekg(fbufpool.beg);
+
+	auto value_of_page = [](int i) {
+		return i * 2 + 1;
+	};
+
+	for (std::size_t i = 0; i < test_num_pages; ++i) {
+		fbufpool.seekg(i * Page::size());
```

Comment:
Why you use std::size_t here but inside `Page::size()` function - uint32_t

Reply:
Good question :smile:. It doesn't make much of a difference since `std::size_t` should be at least the size of `uint32_t` thus not making and weird problems.